indico/modules/events/registration/fields/sessions.py

```diff
@@ -0,0 +1,59 @@
+# This file is part of Indico.
+# Copyright (C) 2002 - 2024 CERN
+#
+# Indico is free software; you can redistribute it and/or
+# modify it under the terms of the MIT License; see the
+# LICENSE file for more details.
+
+from marshmallow import ValidationError, fields, validate
+
+from indico.core.db.sqlalchemy import db
+from indico.modules.events.registration.fields.base import RegistrationFormFieldBase
+from indico.modules.events.registration.models.registrations import RegistrationData
+from indico.modules.events.sessions.models.blocks import SessionBlock
+from indico.util.i18n import _
+
+
+class SessionsField(RegistrationFormFieldBase):
+    name = 'sessions'
+    mm_field_class = fields.List
+    mm_field_args = (fields.Integer,)
+    setup_schema_fields = {
+        'display': fields.String(load_default='expanded'),
+        'minimum': fields.Integer(load_default=0, validate=validate.Range(0)),
+        'maximum': fields.Integer(load_default=0, validate=validate.Range(0)),
+    }
+
+    @property
+    def default_value(self):
+        return []
+
+    @property
+    def filter_choices(self):
+        sessions_ids = [s.id for s in self.form_item.registration_form.event.sessions]
+        return {str(s.id): s.full_title for s in SessionBlock.query.filter(SessionBlock.session_id.in_(sessions_ids))}
+
+    def get_validators(self, existing_registration):
```

Comment:
You also need to validate in here that any session ID passed to this field is either:

- a valid session block (which the user can access) for the event
- in the existing_registration's data (because when deleting a session block it's not removed from registration data, so of course we do not want to break updating a registration after that point)

Reply:
I have added two extra validators: One to check the data integrity (it's a list of integers) and another to check the user can access all SessionBlocks those session blocks, I think it's enough, but let me know if you find anything else needed.